File: src/algorithm_java/leetcode/Problem152.java
package algorithm.leetcode;

public class Problem152 {

    public static int maxProduct(int[] nums) {
        int[] maxDp = new int[nums.length];
        int[] minDp = new int[nums.length];

        maxDp[0] = nums[0];
        minDp[0] = nums[0];
        int maxProduct = nums[0];
        for (int i = 1; i < nums.length; i++) {
            maxDp[i] = Math.max(Math.max(maxDp[i - 1] * nums[i], minDp[i - 1] * nums[i]), nums[i]);
            minDp[i] = Math.min(Math.min(minDp[i - 1] * nums[i], maxDp[i - 1] * nums[i]), nums[i]);
            if (maxDp[i] > maxProduct)
                maxProduct = maxDp[i];
        }
        return maxProduct;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{2,3,-2,4};
        System.out.println(maxProduct(nums));
    }
}

File: src/algorithm_cpp/problem152.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
	vector<int> dpmax(nums.size(), 0);
	vector<int> dpmin(nums.size(), 0);
	dpmax[0] = nums[0];
	dpmin[0] = nums[0];
	int rt = nums[0];
	for (int i = 1; i < nums.size(); i++) {
	    if (nums[i] > 0) {
		dpmax[i] = max(nums[i], dpmax[i - 1] * nums[i]);
		dpmin[i] = min(nums[i], dpmin[i - 1] * nums[i]);
	    } 
	    if (nums[i] < 0) {
		dpmax[i] = max(nums[i], dpmin[i - 1] * nums[i]);
		dpmin[i] = min(nums[i], dpmax[i - 1] * nums[i]);
	    }
	    if (dpmax[i] > rt) rt = dpmax[i];
	}
	return rt;
    }
};
